zerver/actions/message_flags.py

```diff
@@ -290,27 +291,37 @@ def do_update_message_flags(
                 if message_id not in message_ids_in_unsubscribed_streams
             ]
 
-        query = UserMessage.select_for_update_query().filter(
-            user_profile=user_profile, message_id__in=messages
-        )
-
-        um_message_ids = {um.message_id for um in query}
-        if flag == "read" and is_adding:
+        ums = {
+            um.message_id: um
+            for um in UserMessage.select_for_update_query().filter(
+                user_profile=user_profile, message_id__in=messages
+            )
+        }
+
+        messages = [
+            message_id
+            for message_id in messages
+            if (int(ums[message_id].flags) if message_id in ums else DEFAULT_HISTORICAL_FLAGS)
```

Comment:
I found this line pretty hard to read with how the line-wrapping works; not sure I know how to fix it, but it does look correct.